cpp/src/parquet/column_reader.h

```diff
@@ -302,8 +303,274 @@ class TypedColumnReader : public ColumnReader {
+      ++it2;
+    }
+
+    return result;
+  }
+
+  static RowRanges intersection(const RowRanges& left, const RowRanges& right) {
+    RowRanges result;
+
+    size_t rightIndex = 0;
+    for (const Range& l : left.ranges) {
+      for (size_t i = rightIndex, n = right.ranges.size(); i < n; ++i) {
+        const Range& r = right.ranges[i];
+        if (l.isBefore(r)) {
+          break;
+        } else if (l.isAfter(r)) {
+          rightIndex = i + 1;
+          continue;
+        }
+        result.add(Range::intersection(l, r));
+      }
+    }
+
+    return result;
+  }
+
+  RowRanges slice(const int64_t from, const int64_t to) const {
+    RowRanges result;
+    for (const Range& range : ranges) {
+      if (range.from >= from && range.to <= to) {
+        result.add(range);
+      }
+    }
+    return result;
+  }
+
+  void add(const Range& range, bool merge = true) {
+    Range rangeToAdd = range;
+    if (merge) {
+      for (int i = static_cast<int>(ranges.size()) - 1; i >= 0; --i) {
+        Range last = ranges[i];
+        if (last.isAfter(range)) {
+          throw ParquetException(range.toString() + " cannot be added to " +
+                                 this->toString());
+        }
+        const Range u = Range::unionRange(last, rangeToAdd);
+        if (u.from == -1 && u.to == -1) {
+          break;
+        }
+        rangeToAdd = u;
+        ranges.erase(ranges.begin() + i);
+      }
+    } else {
+      if (ranges.size() > 1) assert(rangeToAdd.from > ranges.back().to);
+    }
+    ranges.push_back(rangeToAdd);
+  }
+
+  size_t rowCount() const {
+    size_t cnt = 0;
+    for (const Range& range : ranges) {
+      cnt += range.count();
+    }
+    return cnt;
+  }
+
+  bool isValid() const {
+    if (ranges.size() == 0) {
+      return false;
+    }
+    if (ranges[0].from < 0) {
+      return false;
+    }
+    for (size_t i = 1; i < ranges.size(); i++) {
+      if (ranges[i].from <= ranges[i - 1].to) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool isOverlapping(int64_t from, int64_t to) const {
+    const Range searchRange(from, to);
+    return isOverlapping(searchRange);
+  }
+
+  bool isOverlapping(const Range& searchRange) const {
+    auto it = std::lower_bound(
+        ranges.begin(), ranges.end(), searchRange,
+        [](const Range& r1, const Range& r2) { return r1.isBefore(r2); });
+    return it != ranges.end() && !(*it).isAfter(searchRange);
+  }
+
+  std::vector<Range>& getRanges() { return ranges; }
+
+  const Range& operator[](size_t index) const { return ranges[index]; }
+
+  std::string toString() const {
+    std::string result = "[";
+    for (const Range& range : ranges) {
+      result +=
+          "(" + std::to_string(range.from) + ", " + std::to_string(range.to) + "), ";
+    }
+    if (!ranges.empty()) {
+      result = result.substr(0, result.size() - 2);
+    }
+    result += "]";
+    return result;
+  }
+};
+
+using RowRangesPtr = std::shared_ptr<RowRanges>;
```

Comment:
I don't think we've come to a consensure on these types of acronyms but they are typically discouraged by the C++ style guide.  @pitrou does that match your understanding or is there more up-to-date advice on defining them.

Reply:
Indeed and it's also counter-productive to use a `shared_ptr` around such a tiny struct.

Reply:
I guess the reason of using a shared pointer here is that the row ranges are shared by all column readers.

Reply:
removed such kind of type alias (but still, I don't know why it is not encouraged?)